### ISSUE TITLE
Github actions 6hr time limit

### ISSUE THREAD

**Issue by mahrud** [Macaulay2/M2#1639]
The Github Actions that use autotools and run `make check` are starting to run into Github's 6hr time limit (e.g. [Complexes](https://github.com/Macaulay2/M2/actions/runs/392154407)). There are a few solutions:
- make autotools compile faster (it takes ~1hr longer just to build libraries, Macaulay2, and install packages than CMake)
- [x] don't run the library checks with autotools (really wasteful to run all tests for almost all libraries twice on every push, more than those libraries get tested on their own!)
- run package tests with CMake instead, which can run them in parallel
- finish and merge #1546

All are good solutions, we should probably do all, but we should do at least one very soon.

**Comment by DanGrayson** [comment]
Also check whether we are installing all the Ubuntu packages we could be installing.

**Comment by DanGrayson** [comment]
Good points, all of them.  I've just implemented one of them:
```diff
commit 1e47260755100936bafd60d64013672dedd3941f (HEAD -> development, upstream/development)
Author: Daniel R. Grayson <dan@math.uiuc.edu>
Date:   Mon Nov 30 19:49:40 2020 -0600

    omit checking the libraries

diff --git a/.github/workflows/test_build.yml b/.github/workflows/test_build.yml
index 0f851c497..aae8c12c5 100644
--- a/.github/workflows/test_build.yml
+++ b/.github/workflows/test_build.yml
@@ -182,7 +182,7 @@ jobs:
       - name: Run Tests using Autotools
         if: matrix.build-system == 'autotools' && matrix.os == 'ubuntu-latest'
         run: |
-          make check
+          make check -o check-in-libraries
           make -C Macaulay2/html-check-links check
           make validate-html
 
```

**Comment by mahrud** [comment]
> Also check whether we are installing all the Ubuntu packages we could be installing.

Yes, I think CMake is using all of them + flint from Linuxbew (because the new version isn't on Ubuntu yet), and #1412 is about allowing the same with Autotools. Flint is the biggest bottleneck for Autotools, because factory also depends on it.

**Comment by DanGrayson** [comment]
I just checked: flint is still at 2.5.2 on both Ubuntu 18.04 and 20.04 (to which github will soon switch).  But we should be getting factory from `singular-factory`.

**Comment by d-torrance** [comment]
Another option for Ubuntu packages is using my PPA.  It has bionic (Ubuntu 18.04, which I think is what Github Actions uses) packages for cohomcalg, flint 2.6.3, mpsolve, and topcom, which are the only build build dependencies not in the Ubuntu repositories.

https://launchpad.net/~profzoom/+archive/ubuntu/macaulay2/+index?field.series_filter=bionic

**Comment by DanGrayson** [comment]
That sounds good!

**Comment by d-torrance** [comment]
Even after adding my PPA and installing the mentioned packages, we're still building a bunch of stuff:

```
configure: using BUILDLIBLIST  =  factory frobby gtest
configure: using BUILDSUBLIST  =  memtailor mathic mathicgb
configure: using BUILDPROGLIST =  4ti2 normaliz csdp nauty lrslib gftables
configure: using BUILDLIST     =  factory frobby gtest 4ti2 normaliz csdp nauty lrslib gftables
```
All of these things exist in Ubuntu.  Should I go ahead and add another commit installing these packages too?

**Comment by d-torrance** [comment]
Well, we'll still need to build nauty due to #1408.  (I've submitted a patch to get the problem fixed in Ubuntu 20.04, but it hasn't been sponsored yet.)

**Comment by mahrud** [comment]
> I just checked: flint is still at 2.5.2 on both Ubuntu 18.04 and 20.04 (to which github will soon switch).  But we should be getting factory from `singular-factory`.

I think there's also a new version of factory that's incompatible with the old version of flint. At least for getting good performance.

**Comment by DanGrayson** [comment]
> Even after adding my PPA and installing the mentioned packages, we're still building a bunch of stuff:
> 
> ```
> configure: using BUILDLIBLIST  =  factory frobby gtest
> configure: using BUILDSUBLIST  =  memtailor mathic mathicgb
> configure: using BUILDPROGLIST =  4ti2 normaliz csdp nauty lrslib gftables
> configure: using BUILDLIST     =  factory frobby gtest 4ti2 normaliz csdp nauty lrslib gftables
> ```
> 
> All of these things exist in Ubuntu. Should I go ahead and add another commit installing these packages too?

Sounds good to me.

**Comment by DanGrayson** [comment]
8 builds is too many, too.  Which ones should we stop doing?

**Comment by mahrud** [comment]
> 8 builds is too many, too.  Which ones should we stop doing?

The number doesn't matter, if they finish fast, so none. The CMake builds are pretty efficient and test major compilers on both Ubuntu Ave Mac. Currently autotools doesn't work with clang9, and gets a weird bug on Mac with gcc.

**Comment by d-torrance** [comment]
Now we're down to just two things to build!

```
configure: using BUILDLIBLIST  =  gtest
configure: using BUILDSUBLIST  = 
configure: using BUILDPROGLIST =  nauty
configure: using BUILDLIST     =  gtest nauty
```
There's no reason to build gtest if we're not building mathicgb, so we could also skip that, too.  I'll add a check for that later.

**Comment by mikestillman** [comment]
We do use gtest for unit tests for the engine

**Comment by d-torrance** [comment]
> We do use gtest for unit tests for the engine

Yes, but version 1.8.0 (the one in Ubuntu 18.04) is sufficient.  The only reason we're requiring 1.10 is for mathicgb.

**Comment by DanGrayson** [comment]
> > 8 builds is too many, too.  Which ones should we stop doing?
> 
> The number doesn't matter, if they finish fast, so none. The CMake builds are pretty efficient and test major compilers on both Ubuntu and Mac. Currently autotools doesn't work with clang9, and gets a weird bug on Mac with gcc.

Every build succeeded this morning: https://github.com/Macaulay2/M2/actions/runs/393593281

**Comment by DanGrayson** [comment]
> > 8 builds is too many, too.  Which ones should we stop doing?
> 
> The number doesn't matter, if they finish fast, so none. The CMake builds are pretty efficient and test major compilers on both Ubuntu and Mac. Currently autotools doesn't work with clang9, and gets a weird bug on Mac with gcc.

I think we don't have to test clang and gcc9 on two architectures: one would be enough.  So we can get rid of ubuntu-latest-autotools-gcc9, ubuntu-latest-cmake-clang10.

**Comment by DanGrayson** [comment]
I think we also don't have to trigger builds when pushes occur to development and master -- it's too much, and we build daily on those branches, anyway.

**Comment by mahrud** [comment]
> Every build succeeded this morning: https://github.com/Macaulay2/M2/actions/runs/393593281

Note that there are no autotools builds with clang10, and keep in mind the bugs with gcc on mac: #1350 #1357

> I think we don't have to test clang and gcc9 on two architectures: one would be enough. So we can get rid of ubuntu-latest-autotools-gcc9, ubuntu-latest-cmake-clang10.

I lean towards keeping clang10 one. There have been a number of issues and, again, it's pretty fast!

> I think we also don't have to trigger builds when pushes occur to development and master -- it's too much, and we build daily on those branches, anyway.

You and Mike sometimes push directly to those branches, and it doesn't hurt to test that there are no bugs after a merge (this has happened, see #1523), but we are straying from the point of this issue, which is to make the test builds faster and more efficient, not necessarily reducing their number.

**Comment by mahrud** [comment]
@DanGrayson can we drop the html validation test? It takes more than an hour on github! Here are two different runs:
```
2020-12-19T04:27:44.0049889Z make: Entering directory '/home/runner/work/M2/M2/M2/BUILD/build'
2020-12-19T04:27:44.0060413Z -- validating all html and xhtml files in /home/runner/work/M2/M2/M2/BUILD/build/usr-dist/common/share/doc/Macaulay2
2020-12-19T05:43:24.1347719Z 8900 HTML files checked; 0 invalid
2020-12-19T05:43:24.1350263Z make: Leaving directory '/home/runner/work/M2/M2/M2/BUILD/build'
```
```
2020-12-18T09:11:58.3287695Z make: Entering directory '/home/runner/work/M2/M2/M2/BUILD/build'
2020-12-18T09:11:58.3297484Z -- validating all html and xhtml files in /home/runner/work/M2/M2/M2/BUILD/build/usr-dist/common/share/doc/Macaulay2
2020-12-18T10:18:27.8112785Z 8891 HTML files checked; 0 invalid
2020-12-18T10:18:27.8117850Z make: Leaving directory '/home/runner/work/M2/M2/M2/BUILD/build'
```
Something is very inefficient here.

**Comment by DanGrayson** [comment]
Maybe for extensive tests, such as html verification, we could set up a github action that would trigger once a week.

**Comment by mahrud** [comment]
That seems reasonable, but also, why is it necessary? I'm already validating every generated html:
https://github.com/Macaulay2/M2/blob/0205d20ab465e8cb7721cf507e3febf9b580419d/M2/Macaulay2/m2/installPackage.m2#L430
Also, why is it taking so long?

**Comment by DanGrayson** [comment]
It's necessary, because my M2 function "validate" is not as good.  I don't know why it takes so long.

**Comment by mahrud** [comment]
Here's a part of this daily [autotools-macos-latest-default](https://github.com/Macaulay2/M2/runs/1583791062?check_suite_focus=true) build:
```
configure: using BUILDLIBLIST  =  readline factory mpsolve frobby cddlib gtest
configure: using BUILDSUBLIST  =  memtailor mathic mathicgb fflas_ffpack givaro
configure: using BUILDPROGLIST =  4ti2 gfan normaliz csdp nauty lrslib gftables topcom cohomcalg
configure: using BUILDLIST     =  readline factory mpsolve frobby cddlib gtest 4ti2 gfan normaliz csdp nauty lrslib gftables topcom cohomcalg
```
That build took 2 hours and 50 minutes, about an hour of which was building the libraries. Compare with the CMake build with the same parameters, which took 1h 33m and built pretty much the same set of programs in 12 minutes!
```
       Libraries       = factory;fflas_ffpack;frobby;givaro;googletest;mathic;mathicgb;memtailor;mpsolve
       Programs        = 4ti2;cohomcalg;csdp;gfan;lrslib;nauty;normaliz;topcom
```

**Comment by mahrud** [comment]
Another issue: the autotools-macos-latest-default is suspiciously slow. In a recent build, it took 2h 47m, while autotools-ubuntu-latest-default that also runs all the tests only took 14 more minutes. For another comparison, the cmake-macos-latest-default build took only 1h 30m!

Given that Github allows 20 concurrent jobs, but only 5 concurrent macOS jobs (see [here](https://docs.github.com/en/free-pro-team@latest/actions/reference/usage-limits-billing-and-administration#usage-limits)), I think reducing the time that autotools-macos-latest-default takes should be a priority within this issue.

**Comment by DanGrayson** [comment]
> Another issue: the autotools-macos-latest-default is suspiciously slow. In a recent build, it took 2h 47m, while autotools-ubuntu-latest-default that also runs all the tests only took 14 more minutes. For another comparison, the cmake-macos-latest-default build took only 1h 30m!
> 

Here at home it takes me all day to build on the Mac, but in an Ubuntu VM it's just a few hours.  I don't know why.

**Comment by mahrud** [comment]
> Here at home it takes me all day to build on the Mac, but in an Ubuntu VM it's just a few hours. I don't know why.

You should use CMake more, it'll make your life easier! (You should also install ccache for instantaneous rebuilds). Would be good to get M2 into brew, then we can drop the autotools build on macos.

**Comment by mahrud** [comment]
@DanGrayson the autotools-macOS build is in bad shape:
https://github.com/Macaulay2/M2/runs/1657248094
After 2h 50m, this build hasn't even finished building in the e directory.

**Comment by DanGrayson** [comment]
Strange.  Was it cancelled by you?

**Comment by mahrud** [comment]
Yes, I was more interested in seeing the result of a different build.

Related to #1412, autotools on macOS doesn't detect dependencies installed via brew:
```
configure: using BUILDLIST     =  readline factory mpsolve frobby cddlib gtest 4ti2 gfan normaliz csdp nauty lrslib gftables topcom cohomcalg
```
On CMake this is as easy as `` -DCMAKE_SYSTEM_PREFIX_PATH=`brew --prefix` ``, which instructs CMake to search under that prefix for `lib, include, bin, ...` directories. Note that the fix is not hard coding a path; rather, the local brew prefix is queried so that it can be used both on mac and linux.

**Comment by mahrud** [comment]
Note that I had a fix for this in 49341303950ed6b742f51bdc230914044f00cca8, but it wasn't accepted at the time. Feel free to cherry-pick or use it now if you'd like.

**Comment by mahrud** [comment]
Do we really need to build the documentation on every build? I imagine once M2 is compiled on each machine, we can just install packages and run tests 4 builds: {cmake, autotools} x {ubuntu, macos}. Pick a random compiler for each.

**Comment by DanGrayson** [comment]
Running the documentation examples is a good test.  Why not stop running the cmake build with 3 compilers?  We could switch to testing non-default compilers on a schedule, once every week or two.

**Comment by mahrud** [comment]
> Running the documentation examples is a good test.

It's a lengthy test and running them on two systems at most should suffice.

> Why not stop running the cmake build with 3 compilers?

Because autotools works with a limited range of compilers (e.g. doesn't work with clang at all, or older gcc versions), so when people want to compile with something else they need to use CMake. They need to be checked so that compilation issues are detected and people don't have to message me to debug compilation issues.

>  We could switch to testing non-default compilers on a schedule, once every week or two.

This is really pointless because CMake compiles Macaulay2 in ~10min. You can check that M2 compiles and run the engine tests on 6 builds in <15min, then wait 2 hours for two builds to install all packages and run the checks.

**Comment by d-torrance** [comment]
> Because autotools works with a limited range of compilers (e.g. doesn't work with clang at all, ...

It works for me in Ubuntu 20.10:

```
dtorrance@nimmie:~/src/macaulay2/M2/M2/BUILD/doug$ ../../configure CC=clang CXX=clang++

...

dtorrance@nimmie:~/src/macaulay2/M2/M2/BUILD/doug$ make

...

dtorrance@nimmie:~/src/macaulay2/M2/M2/BUILD/doug$  usr-dist/x86_64-Linux-Ubuntu-20.10/bin/M2
Macaulay2, version 1.17.2.1
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems,
               LLLBases, MinimalPrimes, PrimaryDecomposition, ReesAlgebra,
               Saturation, TangentCone
--warning: file doc.css not installed yet in package Style
--warning: file katex not installed yet in package Style
-- *** warning: file needed for URL not found: /home/dtorrance/src/macaulay2/M2/M2/BUILD/doug/usr-dist/common/share/Macaulay2/Style/GoldStar.png

i1 : version#"compiler"

o1 = clang 11.0.0 
```

**Comment by mahrud** [comment]
Sorry, I should have clarified: this is an issue on MacOS, hence why I insisted on having one clang 10 build on macos.

**Comment by d-torrance** [comment]
> Sorry, I should have clarified: this is an issue on MacOS, hence why I insisted on having one clang 10 build on macos.

Ah, gotcha.
